File: csharp/IOBenchmark.cs
using System;
using System.IO;
using System.Threading.Tasks;
using System.Diagnostics;

namespace Benchmark
{
    public class IOBenchmark
    {
        private static Random random = new Random();

        static async Task CreateTestFiles(int count, string taskId)
        {
            var random = new Random();
            var tasks = new Task[count];
            for (int i = 0; i < count; i++)
            {
                int index = i;
                tasks[i] = Task.Run(async () =>
                {
                    var content = $"Test file {index} content with some data: {random.NextDouble()}";
                    var filename = $"temp_{taskId}_{index}.txt";
                    await File.WriteAllTextAsync(filename, content);
                });
            }
            await Task.WhenAll(tasks);
        }

        static async Task<string[]> ReadTestFiles(int count, string taskId)
        {
            var tasks = new Task<string>[count];
            for (int i = 0; i < count; i++)
            {
                int index = i;
                var filename = $"temp_{taskId}_{index}.txt";
                tasks[i] = Task.Run(async () =>
                {
                    // リトライロジック
                    for (int retry = 0; retry < 10; retry++)
                    {
                        try
                        {
                            if (File.Exists(filename))
                            {
                                return await File.ReadAllTextAsync(filename);
                            }
                            await Task.Delay(10);
                        }
                        catch (IOException)
                        {
                            if (retry == 9) throw;
                            await Task.Delay(10);
                        }
                    }
                    throw new FileNotFoundException($"File not found: {filename}");
                });
            }
            return await Task.WhenAll(tasks);
        }

        static async Task CleanupTestFiles(int count, string taskId)
        {
            for (int i = 0; i < count; i++)
            {
                try
                {
                    var filename = $"temp_{taskId}_{i}.txt";
                    if (File.Exists(filename))
                        File.Delete(filename);
                }
                catch { /* ignore */ }
            }
            await Task.CompletedTask;
        }

        static async Task SimulateNetworkDelay(int ms)
        {
            await Task.Delay(ms);
        }

        static async Task<int> IOIntensiveTask(string taskId)
        {
            int fileCount = 50;
            int networkCalls = 20;

            await CreateTestFiles(fileCount, taskId);

            var networkTasks = new Task[networkCalls];
            for (int i = 0; i < networkCalls; i++)
                networkTasks[i] = SimulateNetworkDelay(10 + (int)(new Random().NextDouble() * 20));

            var fileTask = ReadTestFiles(fileCount, taskId);

            await Task.WhenAll(fileTask, Task.WhenAll(networkTasks));

            await CleanupTestFiles(fileCount, taskId);
            return fileTask.Result.Length;
        }

        public static async Task RunIOBenchmark(int taskCount)
        {
            Console.WriteLine("C# I/O benchmark start");

            var start = DateTime.Now;

            var tasks = new Task[taskCount];
            for (int i = 0; i < taskCount; i++)
            {
                int taskIndex = i; // Capture for closure
                tasks[i] = Task.Run(async () =>
                {
                    try
                    {
                        await IOIntensiveTask($"task{taskIndex}");
                        Console.WriteLine($"Task {taskIndex + 1} done");
                    }
                    catch (Exception e)
                    {
                        Console.WriteLine($"Task {taskIndex + 1} failed: {e.Message}");
                    }
                });
            }

            await Task.WhenAll(tasks);

            var elapsed = DateTime.Now - start;
            Console.WriteLine($"All {taskCount} tasks done in {elapsed.TotalSeconds:F4}s");
        }
    }
}

File: node/io-benchmark.js
// I/O-bound benchmark module (standard lightweight version)
const fs = require('fs').promises;
const { performance } = require('perf_hooks');

async function createTestFiles(count, taskId) {
  const promises = [];
  for (let i = 0; i < count; i++) {
    const content = `Test file ${i} content with some data: ${Math.random()}`;
    promises.push(fs.writeFile(`temp_${taskId}_${i}.txt`, content));
  }
  await Promise.all(promises);
}

async function readTestFiles(count, taskId) {
  const promises = [];
  for (let i = 0; i < count; i++) {
    // リトライロジック付き読み取り
    promises.push(
      (async () => {
        for (let retry = 0; retry < 10; retry++) {
          try {
            return await fs.readFile(`temp_${taskId}_${i}.txt`, 'utf8');
          } catch (err) {
            if (retry === 9) throw err;
            await new Promise(resolve => setTimeout(resolve, 10));
          }
        }
      })()
    );
  }
  return await Promise.all(promises);
}

async function cleanupTestFiles(count, taskId) {
  const promises = [];
  for (let i = 0; i < count; i++) {
    promises.push(fs.unlink(`temp_${taskId}_${i}.txt`).catch(() => {}));
  }
  await Promise.all(promises);
}

async function simulateNetworkDelay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function ioIntensiveTask(taskId) {
  const fileCount = 50;
  const networkCalls = 20;
  
  await createTestFiles(fileCount, taskId);
  
  const networkPromises = [];
  for (let i = 0; i < networkCalls; i++) {
    networkPromises.push(simulateNetworkDelay(10 + Math.random() * 20));
  }
  
  const [fileContents] = await Promise.all([
    readTestFiles(fileCount, taskId),
    Promise.all(networkPromises)
  ]);
  
  await cleanupTestFiles(fileCount, taskId);
  return fileContents.length;
}

async function runIOBenchmark(taskCount = 10) {
  console.log("Node.js I/O benchmark start");
  
  const start = performance.now();
  
  const promises = [];
  for (let i = 0; i < taskCount; i++) {
    promises.push(ioIntensiveTask(`task${i}`).then(result => {
      console.log(`Task ${i + 1} done`);
      return result;
    }).catch(err => {
      console.log(`Task ${i + 1} failed: ${err.message}`);
      throw err;
    }));
  }
  
  const results = await Promise.all(promises);
  
  const end = performance.now();
  const elapsed = Math.round(end - start);
  console.log(`All ${taskCount} tasks done in ${elapsed} ms`);
  return { taskCount, elapsed, results: results.length };
}

module.exports = { runIOBenchmark };